zeebe/atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/HeartbeatHandler.java

```diff
@@ -0,0 +1,130 @@
+        receivedHeartbeat = true;
+        ctx.writeAndFlush(createHeartbeat(request.id()));
+        if (!Arrays.equals(request.payload(), HEARTBEAT_PAYLOAD)) {
+          log.warn(
+              "Received unexpected heartbeat payload from {}, perhaps the message subject {} is accidentally reused",
+              request.sender(),
+              HEARTBEAT_SUBJECT);
+        }
+      }
+
+      // Pass the message to the next handler, even though it's a heartbeat.
+      ctx.fireChannelRead(msg);
+    }
+
+    @Override
+    public void userEventTriggered(final ChannelHandlerContext ctx, final Object evt) {
+      if (!(evt instanceof final IdleStateEvent idleStateEvent)) {
+        return;
+      }
+      if (idleStateEvent.state() == IdleState.READER_IDLE) {
+        // don't close the connection if we did not receive any heartbeat
+        if (receivedHeartbeat) {
+          log.warn("Connection {} timed out on the server, closing channel", ctx.channel());
+          ctx.close();
+        }
+      }
+    }
+
+    private ProtocolMessage createHeartbeat(final long id) {
+      return new ProtocolReply(id, HEARTBEAT_PAYLOAD, Status.OK);
+    }
+  }
+
+  static final class Client extends HeartbeatHandler {
+    private final LongHashSet outstandingHeartbeats = new LongHashSet();
+    private final AtomicLong messageIdGenerator;
+    private final Address advertisedAddress;
+
+    Client(final Logger log, final AtomicLong messageIdGenerator, final Address advertisedAddress) {
+      super(log);
+      this.messageIdGenerator = messageIdGenerator;
+      this.advertisedAddress = advertisedAddress;
+    }
+
+    @Override
+    public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
+      if (msg instanceof final ProtocolReply reply && outstandingHeartbeats.contains(reply.id())) {
+        // remove all heartbeats sent before the last update
+        outstandingHeartbeats.removeIfLong(id -> id <= reply.id());
+        if (reply.status() != Status.OK) {
+          log.warn("Received a Heartbeat response with status {}", reply.status());
+        } else if (!Arrays.equals(reply.payload(), HEARTBEAT_PAYLOAD)) {
+          log.warn(
+              "Received unexpected heartbeat payload, perhaps the message subject {} is accidentally reused: {}",
+              HEARTBEAT_SUBJECT,
+              reply);
+        }
+      }
+
+      // Pass the message to the next handler, even though it's a heartbeat.
+      ctx.fireChannelRead(msg);
+    }
+
+    @Override
+    public void userEventTriggered(final ChannelHandlerContext ctx, final Object evt) {
+      if (!(evt instanceof final IdleStateEvent idleStateEvent)) {
+        return;
+      }
+      switch (idleStateEvent.state()) {
+        case READER_IDLE -> {
+          // don't close the connection if we did not send any heartbeat
+          if (!outstandingHeartbeats.isEmpty()) {
+            log.warn("Connection {} timed out on the client, closing channel", ctx.channel());
```

Comment:
:wrench: `"Connection {} timed out on the client after not receiving a heartbeat response from the server in {} ({} heartbeats pending)"`, ctx.channel(), timeout, outstandingHeartbeats.size())"`

For timeout, ideally I would know what's the oldest heartbeat I sent. But it's OK if that's particularly annoying info to get, we can drop that.